docs/tutorial/streamer_design.md

```diff
@@ -1,29 +1,398 @@
-  - Describe the main blocks
-  - Describe what each module does
-- Show the configuration file and demonstrate examples on different configurations leading to different generated architectures
-  - Add some thinking exercises for this!
-- Timing diagrams at the end
 
+# Why Do We Need a Streamer?
 
-# Motivation for Streamers
+Accelerators attain peak performance when data continuously streams into them; otherwise, they spend cycles waiting for data availability. Delays often occur due to mismatched data arrangements in shared memory or congestion among accelerators accessing the same banks concurrently.
 
+It's crucial to differentiate between the data layout in memory and the access pattern of an accelerator. The figure below shows two different memory layouts and how an accelerator would get the data. The memory address on the top-right corner is a guide to show the adresses of each data element. Assume each column represents a separate memory bank and each block signifies a data element, with each bank having only one read and one write access port.
 
-# Affine Address Generation
+![image](https://github.com/KULeuven-MICAS/snitch_cluster/assets/26665295/4428f8d7-2d35-4605-8bec-92929d195643)
 
+The data layout refers to the arrangement of data contents in memory, while the access pattern pertains to how accelerators retrieve data from memory, such as contiguous or strided access. On the left of the figure, the data layout in memory organizes the inputs and outputs on each bank. The accelerator's data access needs to be configured to access data continuously with appropriate stride memory address stride pointing to the memory addresses.
 
-# Detailed Block Diagram of Streamers
+For example, if the data is arranged in memory layout 1, then input A needs to get data in addresses `[0, 4, 8, 12]` and in the exact order. That means the starting `base_address_a=0` and skip counts with `temporal_stride_a=4`. The `target_address_a` is computed in a simple loop:
 
+```
+for(i = 0; i < 4; i++):
+  target_address_a = base_address_a + temporal_stride_a*i;
+```
 
-# Configuring the Generated Streamer Architecture
+Another arrangement in memory layout 2 shows that the data can be arranged in a contiguous fashion. This time, input A needs to access the data in the address sequence `[0,1,2,3]`. Here we set `base_address_a=0` and `temporal_stride_a=1`.
 
-## Exercises
+A more complicated example is when a streamer can get multiple data in parallel. This necessitates that we need to also have an address generation that can do spatial parallelism. Particularly it would be convenient to provide a base address and a stride but have all other ports automatically increment per accelerator port. The figure below shows an example accelerator that takes in 3 inputs in parallel and also produces 3 outputs in parallel:
 
-# Timing Diagrams
+![image](https://github.com/KULeuven-MICAS/snitch_cluster/assets/26665295/2ecaf54a-ae62-42ab-be55-b1cd3c02c377)
```

Comment:
Now there are multiple memory layouts on this page named layout 1, maybe consider changing to data layout 3..5?

Reply:
MARK TODO!